extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/QuarkusTransaction.java

```diff
@@ -0,0 +1,185 @@
+package io.quarkus.narayana.jta;
+
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+import javax.transaction.HeuristicMixedException;
+import javax.transaction.HeuristicRollbackException;
+import javax.transaction.NotSupportedException;
+import javax.transaction.RollbackException;
+import javax.transaction.SystemException;
+import javax.transaction.UserTransaction;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.narayana.jta.runtime.TransactionManagerConfiguration;
+
+/**
+ * A request scoped representation of a transaction.
+ * <p>
+ * If the transaction is not committed it will be automatically rolled back when the request scope is destroyed.
+ */
+@RequestScoped
+public class QuarkusTransaction {
+
+    private static final Logger log = Logger.getLogger(QuarkusTransaction.class);
+
+    private final UserTransaction userTransaction;
+    private final TransactionManagerConfiguration transactionManagerConfiguration;
+    boolean active;
+    boolean rollbackOnly;
+    boolean autoCommit;
+
+    public static QuarkusTransaction instance() {
+        return Arc.container().instance(QuarkusTransaction.class).get();
+    }
+
+    @Inject
+    public QuarkusTransaction(UserTransaction userTransaction,
+            TransactionManagerConfiguration transactionManagerConfiguration) {
+        this.userTransaction = userTransaction;
+        this.transactionManagerConfiguration = transactionManagerConfiguration;
+    }
+
+    public QuarkusTransaction() {
+        //for proxiability
+        this.userTransaction = null;
+        this.transactionManagerConfiguration = null;
+    }
+
+    public void begin() {
+        begin(false, -1);
+    }
+
+    public void beginRequestScoped() {
+        begin(true, -1);
+    }
+
+    public void begin(int timeoutSeconds) {
+        begin(false, timeoutSeconds);
+    }
+
+    public void beginRequestScoped(int timeoutSeconds) {
+        begin(true, timeoutSeconds);
+    }
+
+    private void begin(boolean autoCommit, int timeout) {
+        try {
+            if (active) {
+                throw new RuntimeException("Transaction already active");
+            }
+            this.autoCommit = autoCommit;
+            if (timeout > 0) {
+                userTransaction.setTransactionTimeout(timeout);
+            }
+            active = true;
+            userTransaction.begin();
+            if (timeout > 0) {
+                userTransaction.setTransactionTimeout(
+                        (int) transactionManagerConfiguration.defaultTransactionTimeout.get().toSeconds());
+            }
+        } catch (NotSupportedException | SystemException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void runInTransaction(Runnable r) {
+        runInTransaction(-1, r);
+    }
+
+    public void runInTransaction(int timeoutSeconds, Runnable r) {
```

Comment:
I am wondering if it would it make sense to go for an experience like: 

`inTransaction().run(() -> {...));`
`inTransation().withTimeout(timeout).run(() -> {..});`

Most probably an overkill if the only option is going to be the 'timeout`.